[aot_autograd] handle subclass input mutations correctly in collect_metadata_analysis.py

davidberard98 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #120136
* #120141

This PR fixes the issue in https://github.com/pytorch/pytorch/issues/120188.

In collect_metadata_analysis.py, handling of input/output mutations was different from handling in other locations. In other locations, MUTATED_OUT_GRAPH was used to indicate that mutation would require returning an output; in collect_metadata_analysis.py, any type of mutation was being handled as if it would require returning an output.

This PR changes collect_metadata_analysis to match other callsites and refactors computation of mutation types so that it is a property of the dataclass instead of something that needs to be computed manually when constructing an InputAliasInfo.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @aakhundov @kadeng

Differential Revision: [D53950998](https://our.internmc.facebook.com/intern/diff/D53950998)

davidberard98 committed:
[aot_autograd] support inplace mutations for subclasses v2

[ghstack-poisoned]

davidberard98 committed:
Update on "[aot_autograd] support inplace mutations for subclasses v2"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

davidberard98 committed:
Update on "[aot_autograd] support inplace mutations for subclasses v2"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

davidberard98 commented on torch/_functorch/_aot_autograd/collect_metadata_analysis.py:
```diff
@@ -574,7 +556,7 @@ def view_avoid_dupes_with_primals(t):
             inp
             for inp, info in zip(flat_f_args, input_info)
-            if info.mutates_data or info.mutates_metadata
+            if info.mutation_type == MutationType.MUTATED_OUT_GRAPH
```
@bdhirsh this is the fix for the test case I added in the tests file

bdhirsh commented on torch/_functorch/_aot_autograd/collect_metadata_analysis.py:
```diff
@@ -586,6 +568,9 @@ def view_avoid_dupes_with_primals(t):
         # in the graph outputs list.
+
+        # TODO (dberard) make this use the same logic as schemas.py??
+        # TODO (dberard) why does it think that nothing requires grad after desugaring TwoTensor subclass
```
Your PR looks strictly better (happy to approve) - although agreed that this would be a nice refactor too

davidberard98 replied:
this comment was left over from some debugging, thanks for catching it (and for the review!)

davidberard98 committed:
Update on "[aot_autograd] handle subclass input mutations correctly in collect_metadata_analysis.py"


This PR fixes the issue in https://github.com/pytorch/pytorch/issues/120188.

In collect_metadata_analysis.py, handling of input/output mutations was different from handling in other locations. In other locations, MUTATED_OUT_GRAPH was used to indicate that mutation would require returning an output; in collect_metadata_analysis.py, any type of mutation was being handled as if it would require returning an output.

This PR changes collect_metadata_analysis to match other callsites and refactors computation of mutation types so that it is a property of the dataclass instead of something that needs to be computed manually when constructing an InputAliasInfo.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

davidberard98 commented:
@davidberard98 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D53950998).

davidberard98 commented:
@pytorchbot merge
